### PR TITLE
wip #1441 Trampoline deep operator stacks

### PR DESCRIPTION
**Pull request by simonbasle**

### CODE REVIEW

**Comment by simonbasle** [comment]
@reactor/core-team can I get an initial review with early thoughts? This is an updated draft from something I had started end of last year, and it is taking shape. I'm mainly looking for corner cases and caveats now, so any insight is going to be welcomed!

**Comment by robotmrv** [comment]
@simonbasle   
I have several concerns about this approach: 
1. implicit `Thread` change for operators which should not do this. This contradict with https://github.com/reactor/reactor-core/pull/2123 as if this feature is ON then any operator could change `Thread`, but `2123` also will help with stack trace size as it will help to reduce the number of intermediate operators      
2. It could be undesired behavior in case of `Blocking` code is expected to be executed by the operator or execution should be sequential by applying `.subscribe/publishOn(Schedulers.single())` (something like in https://github.com/rsocket/rsocket-java/pull/811/files#diff-e53bcd153790621b7b00a6444483c8feR231) 
3. it does not solve the problem with inner chains of operators. E.g. there was a defect in spring-framework https://github.com/spring-projects/spring-framework/issues/24652 
this code reproduces similar problem:  
```java
	@Test
	void hugeOperatorChainWithSwitchIfEmpty() {
		Set<String> threadNames = new ConcurrentSkipListSet<>();

		final int bound = 5000;
		Flux<Integer> tooDeep = chain(0, bound, threadNames);

		assertThat(tooDeep.blockLast(Duration.ofSeconds(5))).isEqualTo(bound);
		assertThat(threadNames).startsWith(Thread.currentThread().getName());
		assertThat(threadNames.size()).as("number of threads").isGreaterThan(1);
	}

	Flux<Integer> chain(int i, int bound, Set<String> threadNames) {
		return Flux.defer(() -> i < bound
				? Flux.<Integer>empty().switchIfEmpty(chain(i + 1, bound, threadNames))
				: Flux.just(i).hide()
		).doOnSubscribe(s -> threadNames.add(Thread.currentThread().getName()));
	}
```
this MR does not solve it. I am not sure that case of sequential `5000` `.map()/flatMap()` is real. So does this feature solve real issue?  

Probably `3.` could be resolved if `new Operators.Stacksafe()` will be held in the `ThreadLocal` so child chains would have access to it and reuse it.

File: reactor-core/src/main/java/reactor/core/publisher/Flux.java [hunk 1]
```diff
@@ -8301,6 +8301,7 @@ public final Disposable subscribe(
 	@Override
 	@SuppressWarnings("unchecked")
 	public final void subscribe(Subscriber<? super T> actual) {
+		Operators.Stacksafe stacksafe = new Operators.Stacksafe();
```
Comment (bsideup) [inline reactor-core/src/main/java/reactor/core/publisher/Flux.java:8304]: not related to the PR, but, if we're okay with "1 additional allocation per `subscribe`", I think we can dramatically improve the TCO (and reuse it for the trampolining)

**Comment by bsideup** [comment]
@simonbasle although I get why we need it, this will make a lot harder to answer questions "why I am not getting thread X here".

Also, not really sure how BlockHound & co will behave now. We will definitely need to test it

**Comment by simonbasle** [comment]
@robotmrv thanks for taking a look.

1. I don't think that is entirely true. The operators that are identified as sync would still be there in the chain, so a `doOnEach` hook like sleuth would still see these and skip decorating them. Also, the detected stack depth is reset to 0 if a publishOn or subscribeOn is detected. I need to provide a way for the onEachOperator hook to be applied to the `StackSafeSubscriber` though... 🤔 
2. This is a fair concern, if the publishOn/subscribeOn is not set within `maxStackDepth` of the blocking code
3. This is outside the scope of this issue, I think. Or rather, it is a degenerate case (even without the defer, which _can_ be improved): the switch of publisher defers and splits the subscription chains, defeating the Stacksafe optimization. But the deferred-recursive construction pattern of the chain is probably unsolvable, so yeah this PR would not aim at solving that particular pattern.

File: reactor-core/src/main/java/reactor/core/publisher/Operators.java [hunk 2]
```diff
@@ -1479,7 +1506,190 @@ public Context currentContext() {
...
+			final Scheduler.Worker          worker;
+			final String                    stringRepresentation;
+
+			Subscription subscription;
+
+			StacksafeSubscriber(CoreSubscriber<? super T> downstream,
+					Scheduler.Worker worker, String message) {
+				this.worker = worker;
+				this.downstream = downstream;
+				this.stringRepresentation = "StacksafeSubscriber{" + message + "}";
```
Comment (robotmrv) [inline reactor-core/src/main/java/reactor/core/publisher/Operators.java:1603]: `reactor.core.CoreSubscriber#currentContext` is not taken into account  
need to cache it here

Comment (robotmrv) [reply reactor-core/src/main/java/reactor/core/publisher/Operators.java:1603]:
```java
@Test
void largeOperatorChainSmallerTrampoliningThresholdTrampolines() {
	Set<String> threadNames = new ConcurrentSkipListSet<>();
	Flux<Integer> tooDeep = Flux.deferWithContext(ctx -> Flux.just(0).hide());
	Flux<Integer> tooDeep = Flux.just(0).hide();
	int bound = 7_000;
	for (int i = 0; i <= bound; i++) {
		int currentI = i;
		tooDeep = tooDeep
				.doOnSubscribe(s -> threadNames.add(Thread.currentThread().getName()))
				.map(previous -> currentI);
	}
	assertThat(tooDeep.blockLast(Duration.ofSeconds(5))).isEqualTo(bound);
	assertThat(threadNames).startsWith(Thread.currentThread().getName());
	assertThat(threadNames.size()).as("number of threads").isGreaterThan(1);
}
```
produces 
```
java.lang.StackOverflowError
	at reactor.core.publisher.FluxMap$MapSubscriber.actual(FluxMap.java)
	at reactor.core.publisher.InnerOperator.currentContext(InnerOperator.java:32)
	at reactor.core.publisher.FluxPeek$PeekSubscriber.currentContext(FluxPeek.java:112)
```

**Comment by simonbasle** [comment]
this surfaced too many concerns and more and more subtle corner cases, which tend to indicate such a change would induce subtle bugs and/or undeterministic behavior in user code down the road, so I'll bin it.

**Pull request closed without merging**